src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+
+	/* Fetch the data cells on other shards. For parity update,
+	 * the bitmap is set for the same cells as are replicated.
+	 */
+	rc = agg_fetch_odata_cells(entry, bit_map, cell_cnt,
+				   stripe_ud->asu_recalc);
+	if (rc)
+		goto out;
+
+	if (p > 1) {
+		rc = agg_fetch_remote_parity(entry);
+		if (rc)
+			goto out;
+	}
+
+	if (stripe_ud->asu_recalc)
+		agg_recalc_parity(entry, bit_map, cell_cnt);
+	else
+		rc = agg_update_parity(entry, bit_map, cell_cnt);
+
+out:
+	ABT_eventual_set(stripe_ud->asu_eventual, (void *)&rc, sizeof(rc));
+
+}
+
+/* Driver function for partial stripe update. Fetches the data and then invokes
+ * second function to update the parity.
+ */
+static int
+agg_process_partial_stripe(struct ec_agg_entry *entry)
+{
+	struct ec_agg_stripe_ud	 stripe_ud = { 0 };
+	struct ec_agg_extent	*extent;
+	int			*status;
+	uint8_t			*bit_map = NULL;
+	uint8_t			 fcbit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	uint8_t			 tbit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned long            ss;
+	unsigned int		 i, full_cell_cnt = 0;
+	unsigned int		 cell_cnt = 0;
+	unsigned int		 estart, elen = 0;
+	unsigned int		 eend = 0;
+	int			 rc = 0;
+
+	/* For each contiguous extent, constructable from the extent list,
+	 * determine how many full cells, and how many cells overall,
+	 * are contained in the constructed extent.
+	 */
+	ss = k * len * entry->ae_cur_stripe.as_stripenum;
+	estart = entry->ae_cur_stripe.as_offset;
+	d_list_for_each_entry(extent, &entry->ae_cur_stripe.as_dextents,
 			      ae_link) {
-		daos_epoch_range_t	 epoch_range = { 0 };
-		int			 erc = 0;
-
-		epoch_range.epr_lo = epoch_range.epr_hi = agg_extent->ae_epoch;
-
-		erc = vos_obj_array_remove(agg_param->ap_cont_handle,
-					   entry->ae_oid, &epoch_range,
-					   &entry->ae_dkey, &entry->ae_akey,
-					   &agg_extent->ae_orig_recx);
-		if (erc)
-			D_ERROR("vos_obj_array_remove failed: "DF_RC"\n",
-				DP_RC(erc));
-		if (!rc && erc)
-			rc = erc;
-		ext_counter++;
-		if (ext_counter >= ext_total)
-			break;
+		D_ASSERT(!extent->ae_hole);
+		if (extent->ae_epoch <= entry->ae_par_extent.ape_epoch)
+			continue;
+		if (estart == extent->ae_recx.rx_idx - ss) {
+			eend = estart + extent->ae_recx.rx_nr;
+			elen = extent->ae_recx.rx_nr;
+			continue;
+		}
+		if (extent->ae_recx.rx_idx - ss > eend) {
+			cell_cnt += agg_count_cells(fcbit_map, tbit_map, estart,
+						    elen, k, len, entry->
+						    ae_cur_stripe.as_stripenum,
+						    &full_cell_cnt);
+			estart = extent->ae_recx.rx_idx;
+			elen = 0;
+		}
+		elen += extent->ae_recx.rx_nr;
+		eend += extent->ae_recx.rx_nr;
+	}
+	cell_cnt += agg_count_cells(fcbit_map, tbit_map, estart, elen, k, len,
+				    entry->ae_cur_stripe.as_stripenum,
+				    &full_cell_cnt);
+
+	if (full_cell_cnt >= k / 2 || cell_cnt == k) {
+		stripe_ud.asu_recalc = true;
+		cell_cnt = full_cell_cnt;
+		bit_map = fcbit_map;
+	} else
+		bit_map = tbit_map;
```

Comment:
It looks like to me, we can use these code for both partial update and whole stripe update.  i.e. you do not need tell the full stripe and parital stripe in agg_process_stripe()?

Reply:
The two bitmaps are for full cells vs. partial cells (not stripes).  You need at least half the cells for a stripe totally filled as replicas for parity recalculation to be more efficient than parity update.

Reply:
yes, I mean if full_cell_cnt == k, then you do totally local encoding (agg_encode_local_parity) here.
Anyway, it is just a suggestion.